### PR TITLE
Volcano plot tool

### PR DESCRIPTION
**Pull request by mblue9**
Would like to make a tool for users to make a custom volcano plot. So far have just made a start (more like a skeleton at this point), based on http://www.gettinggeneticsdone.com/2016/01/repel-overlapping-text-labels-in-ggplot2.html

FOR CONTRIBUTOR:
* [ ] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

### CODE REVIEW

File: tools/volcanoplot/volcanoplot.R [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+# Below modified from http://www.gettinggeneticsdone.com/2016/01/repel-overlapping-text-labels-in-ggplot2.html
+
+results <- read.delim(opt$input)
+results$fdr <- results[, opt$fdr_col]
+results$Pvalue <- results[, opt$pval_col]
+results$logFC <- results[, opt$lfc_col]
+results$labels <- results[, opt$label_col]
+
+results <- mutate(results, sig=ifelse((fdr<opt$signif_thresh & logFC>opt$lfc_thresh), "Up", ifelse((fdr<opt$signif_thresh & logFC < -opt$lfc_thresh),"Down", "Not Sig")))
```
Comment (mvdbeek) [inline tools/volcanoplot/volcanoplot.R]: Can we customize "Down" "Up" and "Not Sig" ? I think it'd be nice to be able to for instance highlight GO terms  or entities from another experiment?

File: tools/volcanoplot/volcanoplot.xml [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+]]></command>
+    <inputs>
+        <param name="input" type="data" format="tabular" label="Specify an input file" />
+        <param name="fdr_col" type="data_column" data_ref="input" label="FDR (adjusted P value)" />
+        <param name="pval_col" type="data_column" data_ref="input" label="P value (raw)" />
+        <param name="lfc_col" type="data_column" data_ref="input" label="Log Fold Change" />
+        <param name="label_col" type="data_column" data_ref="input" label="Labels" />
+        <param name="signif_thresh" type="float" value="0.05" label="Significance threshold" help="Default: 0.05"/>
+        <param name="lfc_thresh" type="float" value="0" label="LogFC threshold to colour" help="Default: 0"/>
```
Comment (mvdbeek) [inline tools/volcanoplot/volcanoplot.xml]: If you supply a file in `label_select` this seems to apply to only those points in that file, while you may want to label from a file but still color by log FC, which was surprising to me.

File: tools/volcanoplot/volcanoplot.xml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    #end if
+
+]]></command>
+    <inputs>
+        <param name="input" type="data" format="tabular" label="Specify an input file" />
+        <param name="fdr_col" type="data_column" data_ref="input" label="FDR (adjusted P value)" />
+        <param name="pval_col" type="data_column" data_ref="input" label="P value (raw)" />
+        <param name="lfc_col" type="data_column" data_ref="input" label="Log Fold Change" />
+        <param name="label_col" type="data_column" data_ref="input" label="Labels" />
+        <param name="signif_thresh" type="float" value="0.05" label="Significance threshold" help="Default: 0.05"/>
```
Comment (mvdbeek) [inline tools/volcanoplot/volcanoplot.xml]: I guess we could put a max of 1 here and below

File: tools/volcanoplot/volcanoplot.R (hunk 1 above)
Comment (mblue9) [reply tools/volcanoplot/volcanoplot.R]: Ok I've added in an option for specifying the legend labels, let me know if that works for what you're thinking or not.

File: tools/volcanoplot/volcanoplot.xml (hunk 2 above)
Comment (mblue9) [reply tools/volcanoplot/volcanoplot.xml]: Sorry not sure what you mean here? If you supply a label file then those points are labelled and all points that meet the lfc and adjP thesholds are coloured e.g. like below with 1 gene supplied in the label file

screen shot 2018-10-17 at 4 53 53 pm

File: tools/volcanoplot/volcanoplot.xml (hunk 3 above)
Comment (mblue9) [reply tools/volcanoplot/volcanoplot.xml]: I've added a max of 1 for signif_thresh but not for lfc_thresh as that could be more than 1 (was that what you meant by below)

**Comment by mblue9** [comment]
Thanks for the review @mvdbeek ! If you think I should make more changes just let me know.

File: tools/volcanoplot/volcanoplot.xml (hunk 2 above)
Comment (mvdbeek) [reply tools/volcanoplot/volcanoplot.xml]: Hmm, that didn't work for me, but cool if it works than it's fine.

File: tools/volcanoplot/volcanoplot.xml (hunk 3 above)
Comment (mvdbeek) [reply tools/volcanoplot/volcanoplot.xml]: Yep, thanks

**Comment by mblue9** [comment]
Thanks @mvdbeek !

**Comment by marianaram** [comment]
Hello,
I'm trying to use this tool but the following error appears :
Warning message:
As of rlang 0.4.0, dplyr must be at least version 0.8.0.
✖ dplyr 0.7.8 is too old for rlang 0.4.4.
ℹ Please update dplyr with `install.packages("dplyr")` and restart R. 
Warning messages:
1: The `printer` argument is deprecated as of rla  
Can you help me please ?
Thanks :)

**Comment by joylwq** [comment]
Hello,
I encountered the same error as @marianaram too.
How should we do the update ? Hope to get a reply soon! Thank you very much.

Warning message:
As of rlang 0.4.0, dplyr must be at least version 0.8.0.
✖ dplyr 0.7.8 is too old for rlang 0.4.4.
ℹ Please update dplyr with `install.packages("dplyr")` and restart R. 
Error in getopt(spec) : flag "Y" requires an argument

**Comment by marianaram** [comment]
Hello @joylwq,

The problem comes from the input file. Make sure you replace "," by ".", and "E" by "e". This should work !

**Comment by joylwq** [comment]
Hi @marianaram ,
You are right! Thank you very much!

Hi @bgruening , @mblue9 , and @mvdbeek ,
Thanks for mentioning this pull request. Seems like Deseq2-derived files (or any files for that matter) cannot work if there's "," or "E" present.
Hope someone can fix this since it's common to have "E" in P-value columns in Excel and it is troublesome to edit to "e" for large files and with very small P-values with E^-55 for example.

Thanks everyone for their help! :)

**Comment by mblue9** [comment]
Hi @joylwq and @marianaram apologies, it's taken a while to get a chance to look into this.  I'm updating the tool now here https://github.com/galaxyproject/tools-iuc/pull/3713

I've tried to reproduce the issue you reported with "E" and "," to see if I can fix it, but I can't reproduce yet e.g. an input with this seems to work (plot is produced without error)

| Gene   | log2FoldChange | pvalue | padj
| ------  | ------------ | ----- | ----- |
| DOX,6 | 0.51 | 1.861E-08 | 0.0003053

I know it's been a while but wondering if you have a few rows or a Galaxy history with this issue that I could test with? Or if you have any other requests for the tool let me know and I'll see what I can do.

**Pull request merged**